constructor/signing.py

```diff
@@ -0,0 +1,220 @@
+        )
+
+    def get_signing_command(self) -> str:
+        timestamp_server = os.environ.get(
+            "CONSTRUCTOR_SIGNTOOL_TIMESTAMP_SERVER_URL",
+            "http://timestamp.sectigo.com"
+        )
+        timestamp_digest = os.environ.get(
+            "CONSTRUCTOR_SIGNTOOL_TIMESTAMP_DIGEST",
+            "sha256"
+        )
+        file_digest = os.environ.get(
+            "CONSTRUCTOR_SIGNTOOL_FILE_DIGEST",
+            "sha256"
+        )
+        command = (
+            f"{win_str_esc(self.executable)} sign /f {win_str_esc(self.certificate_file)} "
+            f"/tr {win_str_esc(timestamp_server)} /td {timestamp_digest} /fd {file_digest}"
+        )
+        if "CONSTRUCTOR_PFX_CERTIFICATE_PASSWORD" in os.environ:
+            # signtool can get the password from the env var on its own
+            command += ' /p "%CONSTRUCTOR_PFX_CERTIFICATE_PASSWORD%"'
+        return command
+
+    def verify_signing_tool(self):
+        super()._verify_tool_is_available()
+        if not Path(self.certificate_file).exists():
+            raise FileNotFoundError(f"Could not find certificate file {self.certificate_file}.")
+        check_call([self.executable, "/?"], stdout=PIPE, stderr=PIPE)
+
+    def verify_signature(self, installer_file: Union[str, Path]):
+        proc = run(
+            [self.executable, "verify", "/v", str(installer_file)],
+            stdout=PIPE,
+            stderr=STDOUT,
+            text=True,
+        )
+        logger.info(proc.stdout)
+        if "SignTool Error: No signature found" in proc.stdout:
+            # This is a signing error!
+            proc.check_returncode()
+        elif proc.returncode:
+            # we had errors but maybe not critical ones
+            logger.error(
+                f"SignTool could find a signature in {installer_file} but detected errors. "
+                "This is expected for untrusted (development) certificates. "
+                "If it is supposed to be trusted, please check your certificate!"
+            )
+
+
+class AzureSignTool(SigningTool):
+    def __init__(self):
+        super().__init__(os.environ.get("AZURE_SIGNTOOL_PATH", "AzureSignTool"))
+
+    def get_signing_command(self) -> str:
+
+        required_env_vars = (
+            "AZURE_SIGNTOOL_KEY_VAULT_URL",
+            "AZURE_SIGNTOOL_KEY_VAULT_CERTIFICATE",
+        )
+        check_required_env_vars(required_env_vars)
+        timestamp_server = os.environ.get(
+            "AZURE_SIGNTOOL_TIMESTAMP_SERVER_URL",
+            "http://timestamp.sectigo.com"
+        )
+        timestamp_digest = os.environ.get(
+            "AZURE_SIGNTOOL_TIMESTAMP_DIGEST",
+            "sha256"
+        )
+        file_digest = os.environ.get(
+            "AZURE_SIGNTOOL_FILE_DIGEST",
+            "sha256"
+        )
+
+        command = (
+            f"{self.executable} sign -v"
+            ' -kvu %AZURE_SIGNTOOL_KEY_VAULT_URL%'
+            ' -kvc %AZURE_SIGNTOOL_KEY_VAULT_CERTIFICATE%'
```

Comment:
Should we quote this?

Reply:
We can. I just tested it locally and the test worked fine with quotes, too.